File: web/src/connections/storageUrl.ts
/// What somebody fills in for a bucket, and the URL that comes out of it.
///
/// The studio's storage connections are URLs, which is right for a `WDS_CONN_*` in a Compose file and
/// wrong as a thing to type by hand: nobody remembers whether the account goes in the host or the
/// path, or which of `?key=`, `?sas=` and `?connectionstring=` Azure wants. So the wizard collects
/// fields and this builds the URL — and it lives on its own, without React, because the escaping is
/// the part that has to be right.

export type StorageProvider = "s3" | "azblob" | "gs" | "file";

/// How the connection proves who it is. `identity` means nothing is written down: the managed
/// identity, the instance role, or application default credentials.
export type StorageAuth = "identity" | "keys" | "sas" | "connectionstring" | "json";

export interface StorageDraft {
  provider: StorageProvider;
  /// The bucket for S3 and Google, the container for Azure, the path for a folder.
  container: string;
  /// The storage account. Azure only, and not needed when a connection string carries it.
  account: string;
  /// Opens the connection inside the container rather than at its root.
  prefix: string;
  region: string;
  /// An S3-compatible endpoint: MinIO, R2, Wasabi, Ceph.
  endpoint: string;
  auth: StorageAuth;
  /// S3: the access key. Azure: unused. Google: unused.
  access: string;
  /// S3: the secret key. Azure: the account key. The one field that is a secret in every provider.
  secret: string;
  /// An Azure shared-access signature, or an Azure connection string, or a Google service-account
  /// JSON — whichever `auth` names.
  token: string;
  /// Google HMAC keys, which querying through DuckDB needs and browsing does not.
  hmac: string;
  hmacSecret: string;
}

export const emptyDraft: StorageDraft = {
  provider: "s3", container: "", account: "", prefix: "", region: "", endpoint: "",
  auth: "identity", access: "", secret: "", token: "", hmac: "", hmacSecret: "",
};

/// Which auth choices a provider actually has. Offering an Azure SAS for an S3 bucket would only
/// invite the question.
export function authChoices(provider: StorageProvider): { value: StorageAuth; label: string }[] {
  switch (provider) {
    case "s3":
      return [
        { value: "identity", label: "The machine's own role" },
        { value: "keys", label: "Access key and secret" },
      ];
    case "azblob":
      return [
        { value: "identity", label: "Managed identity" },
        { value: "keys", label: "Account key" },
        { value: "sas", label: "Shared-access signature" },
        { value: "connectionstring", label: "Connection string" },
      ];
    case "gs":
      return [
        { value: "identity", label: "Application default credentials" },
        { value: "json", label: "Service-account JSON" },
      ];
    case "file":
      return [{ value: "identity", label: "The container's own filesystem" }];
  }
}

/// What is missing or wrong, in the order somebody would fix it. Empty means the draft builds.
export function storageProblems(draft: StorageDraft): string[] {
  const problems: string[] = [];

  if (draft.provider === "file") {
    if (!draft.container.trim()) problems.push("a path inside the container is needed");
    return problems;
  }

  if (!draft.container.trim())
    problems.push(draft.provider === "azblob" ? "a container is needed" : "a bucket is needed");

  if (draft.provider === "azblob" && draft.auth !== "connectionstring" && !draft.account.trim())
    problems.push("a storage account is needed");

  if (draft.auth === "keys") {
    if (draft.provider === "s3" && !draft.access.trim()) problems.push("an access key is needed");
    if (!draft.secret.trim())
      problems.push(draft.provider === "s3" ? "a secret key is needed" : "an account key is needed");
  }

  if (draft.auth === "sas" && !draft.token.trim()) problems.push("a shared-access signature is needed");

  if (draft.auth === "connectionstring" && !draft.token.trim())
    problems.push("a connection string is needed");

  if (draft.auth === "json" && !draft.token.trim())
    problems.push("the service-account JSON is needed");

  if (draft.endpoint.trim() && !/^https?:\/\//i.test(draft.endpoint.trim()))
    problems.push("an endpoint starts with http:// or https://");

  // One HMAC half without the other is a query that fails later rather than a mistake caught now.
  if ((draft.hmac.trim() === "") !== (draft.hmacSecret.trim() === ""))
    problems.push("HMAC keys come in pairs: both the key and its secret");

  return problems;
}

/// The URL the studio stores. Secrets travel as escaped query values, which is what keeps a base64
/// key with a `+` in it intact.
export function buildStorageUrl(draft: StorageDraft): string {
  const prefix = draft.prefix.trim().replace(/^\/+|\/+$/g, "");
  const container = draft.container.trim().replace(/^\/+|\/+$/g, "");

  if (draft.provider === "file") {
    // file:///data/incoming — three slashes, and the path is the whole of it.
    const path = draft.container.trim().replace(/\\/g, "/").replace(/^\/+/, "");
    return `file:///${path}${prefix ? `/${prefix}` : ""}`;
  }

  const options: [string, string][] = [];

  if (draft.provider === "s3") {
    if (draft.region.trim()) options.push(["region", draft.region.trim()]);
    if (draft.endpoint.trim()) options.push(["endpoint", draft.endpoint.trim()]);
    if (draft.auth === "keys") {
      options.push(["access", draft.access.trim()]);
      options.push(["secret", draft.secret]);
    }
  }

  if (draft.provider === "azblob") {
    if (draft.endpoint.trim()) options.push(["endpoint", draft.endpoint.trim()]);
    if (draft.auth === "keys") options.push(["key", draft.secret]);
    if (draft.auth === "sas") options.push(["sas", draft.token.replace(/^\?/, "")]);
    if (draft.auth === "connectionstring") options.push(["connectionstring", draft.token.trim()]);
  }

  if (draft.provider === "gs") {
    if (draft.auth === "json") options.push(["credentials", draft.token.trim()]);
    if (draft.hmac.trim()) {
      options.push(["hmac", draft.hmac.trim()]);
      options.push(["hmacsecret", draft.hmacSecret]);
    }
  }

  const query = options.length === 0
    ? ""
    : "?" + options.map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join("&");

  const path = [container, prefix].filter(Boolean).join("/");

  // Azure puts the account in the host and the container in the path; the others put the bucket in
  // the host. An Azure connection string already names the account, so the host may stay empty.
  if (draft.provider === "azblob")
    return `azblob://${draft.account.trim()}/${path}${query}`;

  return `${draft.provider}://${path}${query}`;
}

/// The same URL with every secret replaced, for showing on screen. A wizard that prints an account
/// key in a preview is a wizard that puts it in a screenshot.
export function maskStorageUrl(url: string): string {
  return url.replace(
    /([?&](?:secret|key|sas|connectionstring|credentials|hmacsecret)=)([^&]*)/gi,
    (_, head: string) => `${head}…`);
}

/// A name somebody does not have to invent: LAKE from `s3://lake`, EXPORTS from a container.
export function suggestName(draft: StorageDraft): string {
  const source = draft.provider === "file"
    ? draft.container.trim().replace(/\\/g, "/").split("/").filter(Boolean).pop() ?? ""
    : draft.container.trim();

  return source.replace(/[^A-Za-z0-9]+/g, "_").replace(/^_+|_+$/g, "").toUpperCase();
}

File: src/WebDataStudio.Server/Storage/StorageUrl.cs

namespace WebDataStudio.Server.Storage;

/// Which object store a connection points at.
public enum StorageProvider { S3, AzureBlob, GoogleCloud, Local }

/// A storage connection, taken apart. `Container` is the bucket or the blob container; `Prefix` is
/// the part of the key space this connection is limited to, so a connection can hand somebody one
/// folder of a bucket rather than the whole thing.
public sealed record StorageTarget(
    StorageProvider Provider,
    string? Account,
    string Container,
    string Prefix,
    IReadOnlyDictionary<string, string> Options)
{
    public string? Option(string name) => Options.TryGetValue(name, out var value) ? value : null;

    /// True when the connection carries credentials of its own. Without them the machine's identity
    /// is used — a managed identity on Azure, an instance role on AWS, application default
    /// credentials on Google.
    public bool HasExplicitCredentials =>
        Option("key") is not null || Option("sas") is not null
        || (Option("access") is not null && Option("secret") is not null)
        || (Option("hmac") is not null && Option("hmacsecret") is not null);

    /// The key a prefix-scoped connection actually addresses.
    public string KeyOf(string relative) =>
        Prefix.Length == 0 ? relative.TrimStart('/') : $"{Prefix.TrimEnd('/')}/{relative.TrimStart('/')}";
}

/// Reads the URL form a storage connection is configured with.
///
///   s3://bucket/prefix?region=eu-central-1&amp;endpoint=https://minio:9000&amp;access=…&amp;secret=…
///   azblob://account/container/prefix?key=…            (or ?sas=…)
///   gs://bucket/prefix?hmac=…&amp;hmacsecret=…
///   file:///data/incoming
///
/// One engine id — `storage` — and the scheme decides the provider. Credentials are optional
/// everywhere: without them the machine's own identity is used.
public static class StorageUrl
{
    public static bool IsStorageScheme(string scheme) => Provider(scheme) is not null;

    private static StorageProvider? Provider(string scheme) => scheme.ToLowerInvariant() switch
    {
        "s3" => StorageProvider.S3,
        "azblob" or "azure" => StorageProvider.AzureBlob,
        "gs" or "gcs" => StorageProvider.GoogleCloud,
        "file" => StorageProvider.Local,
        _ => null,
    };

    public static StorageTarget Parse(string url)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri))
            throw new FormatException($"'{url}' is not a storage URL");

        var provider = Provider(uri.Scheme)
            ?? throw new FormatException($"'{uri.Scheme}' is not a storage scheme");

        var options = Options(uri);

        // A file URL is a path, not a host and a container. Everything after the root is the prefix,
        // and the "container" is the root itself so the rest of the code has one shape to work with.
        if (provider == StorageProvider.Local)
        {
            var path = Uri.UnescapeDataString(uri.LocalPath);
            if (path.Length == 0) throw new FormatException("a file:// storage connection needs a path");

            return new StorageTarget(provider, null, path, "", options);
        }

        // s3://bucket/prefix — the host is the bucket. azblob://account/container/prefix — the host
        // is the account and the first path segment is the container, because a storage account has
        // more than one and the connection has to say which.
        var segments = uri.AbsolutePath.Trim('/').Split('/', StringSplitOptions.RemoveEmptyEntries);

        if (provider == StorageProvider.AzureBlob)
        {
            // azblob:///container?connectionstring=... — an app host hands over a connection string
            // or a service URI, and the account name is already inside it. Asking for it a second
            // time in the URL would be a way to write down two different accounts.
            var account = uri.Host.Length > 0 ? uri.Host : null;

            if (account is null && options.GetValueOrDefault("connectionstring") is not { Length: > 0 })
                throw new FormatException(
                    "an azblob:// connection needs an account: azblob://account/container");

            if (segments.Length == 0)
                throw new FormatException("an azblob:// connection needs a container: azblob://account/container");

            return new StorageTarget(provider, account, segments[0],
                string.Join('/', segments.Skip(1)), options);
        }

        if (uri.Host.Length == 0) throw new FormatException($"'{url}' names no bucket");

        return new StorageTarget(provider, null, uri.Host, string.Join('/', segments), options);
    }

    private static Dictionary<string, string> Options(Uri uri)
    {
        var options = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        foreach (var pair in uri.Query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            // Split on the first '=' only: an Azure connection string is itself full of them.
            var split = pair.Split('=', 2);

            // Unescaped as a URI component, not as a query string: a query parser turns '+' into a
            // space, and a base64 account key is full of them — which broke the key silently.
            options[Uri.UnescapeDataString(split[0])] =
                split.Length > 1 ? Uri.UnescapeDataString(split[1]) : "";
        }

        return options;
    }
}
